### ISSUE TITLE
TypedDict type narrowing depends on argument order

### ISSUE THREAD

**Issue by rmorshea** [microsoft/pyright#9649]
**Describe the bug**

Pyright's ability to infer whether a dict literal passed to a generic function is a `TypedDict` depends on argument order. That is, when a dict literal is passed to a generic parameter of a function it't always treated as a `dict` unless there is a prior argument that requires the dict literal to be a `TypedDict`.

**Code or Screenshots**

In the code below ([playground link](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoUSKVATwvSwHMylarQAqPJABMAIujRtw0brwFCqNKOIpIAagRAsWoqAF5NEvSAAUAIlE2AlIZYx4BHDk0ALPvytapWTQHAC4WKAioYDAwEKgcVANDSSRgKIBXLBgAfWB0EASrYEyYOIYmImQAbSrRAF0AGigAOWwkRqh9fjjRBygAWgA%2BFra4vHHnFLTirOy3Qq6eppnS%2BkY3SqQa%2BqbWrHa%2BoZH9sYnk1Kh0nCRs1B8BK0XvX0PhvaRTvGcV3PzCq5ud18TQA3jZomAbHEbER9DYAL59KAAYigYAA1mwSnN3KgrGCIVCoDC4fCmgDbvd%2BEjUUgQOAQNDJHJUFUEiAmuy6jYyJ4sGBaO4cOh%2BFhmEhOGBiaIqTYsbM8gU8V0TASYkSSSAEcsSiYKUCBDSoHSGUyWWzEpzEtzeVB%2BYKPCKxZtJdLZSwgA)) the first usage is valid but the latter are not:

```python
from collections.abc import Callable
from typing import TypedDict
from typing import TypeVar

T = TypeVar("T")


class Thing(TypedDict):
    foo: str


def func_first(func: Callable[[T], None], arg: T) -> None: ...


def func_last(arg: T, func: Callable[[T], None]) -> None: ...


def use_thing(arg: Thing) -> None: ...


func_first(use_thing, {"foo": "bar"})  # ok
func_last({"foo": "bar"}, use_thing)  # error: "dict[str, str]" is not assignable to "Thing"
func_first(arg={"foo": "bar"}, func=use_thing)  # error: "dict[str, str]" is not assignable to "Thing"
```

The last example is particularly confusing since `func` appears before `arg` in the function signature of `func_first` but Pyright appears to prefer evaluating in the order or the given arguments instead. At the very least evaluating in the order of the signature parameters would at least ensure more consistent behavior.

**Comment by erictraut** [comment]
Yes, this is a consequence of the way that pyright's constraint solver works. Arguments are evaluated left to right, and constraints are added as they are discovered. In `func_fist`, the constraint `T <= Thing` is established. This constraint is then used to evaluate the dict literal using bidirectional type inference. When it's done in the other order, the constraint `T >= dict[str, str]` is established.

The typing spec doesn't dictate how type checkers should perform inference or constraint solving. The inference and constraint-solving algorithms in pyright have been tuned over many years to accommodate most common use cases. There will always be cases (like the one in your example) that the algorithm will fail on. That's unfortunately, but it's a reality given the constraints of the problem. Theoretically, a constraint solver could be written that establishes all possible sets of constraints and then heuristically picks from among them, but such a solver would be very slow.

I'm going to close this as "won't fix" because pyright is working as intended here.

To work around this limitation, you can assign the dict literal expression to a temporary variable whose type has been declared as `Thing`.

```python
temp: Thing = {"foo": "bar"}
func_last(temp, use_thing)
```

**Comment by rmorshea** [comment]
Understood. Though, out of curiosity, and to push back a little, what's the reason for choosing argument order vs signature order? It would seem preferable to allow library authors to order parameters in their function signatures to ensure Pyright correctly infers types regardless of how they are called.
